uniffi_bindgen/src/bindings/python/templates/ErrorTemplate.py

```diff
@@ -0,0 +1,51 @@
+class RustError(ctypes.Structure):
+    _fields_ = [
+        ("code", ctypes.c_int32),
+        ("message", ctypes.c_char_p),
+    ]
+
+    def __str__(self):
+        return "RustError(code={}, message={})".format(self.code, str(self.message, "utf-8")) 
+
+{% for e in ci.iter_error_definitions() %}
+class {{e.name()}}:
+    {%- for value in e.values() %}
+    class {{value}}(Exception):
+        pass
+    {%- endfor %}
+
+    @staticmethod
+    def raise_err(err):
+        {%- for value in e.values() %}
+        if err.code == {{loop.index}}:
+            raise {{e.name()}}.{{value}}(err.__str__())
+        {% endfor %}
+        raise Exception("Unknown {{e.name()}} error code")
+{% endfor %}
+
+class RustErrorHelper:
+    def __init__(self):
+        self.err = RustError()
+
+    def __apply__(self, fn):
+        return fn(ctypes.byref(self.err))
+
+    def __raise__(self):
+    {%- for e in ci.iter_error_definitions() %}
+        if self.err.code == {{loop.index}}:
+            return {{e.name()}}.raise_err(self.err)
+    {% endfor %}
+        return
+
+    def __reset__(self):
+        del self.err
+        self.err = RustError()
+
+    def try_raise(self, fn):
+        self.__reset__()
+        result = self.__apply__(fn)
+        self.__raise__()
+        
+        return result
+
+_RustErrorHelper = RustErrorHelper()
```

Comment:
Using a stateful object for this is going to be a threading hazard - imagine two threads calling into two different functions at the same time, but both using `self.err` as their out error parameter. We might accidentally fail to report an error, or report the wrong type of error, if they concurrently modified the shared `self.err` struct.

Fortunately, I don't think you need to use a stateful objectfor this. I suggest using a stateless helper function instead, using python's star-args syntax to make it generic in the number of arguments. Something like:

```
def rust_call_with_err(errorClass, fn, *args):
    err = RustError()
    args = args + (ctypes.byref(self.err),)  # `args` is a tuple, this puts `err` on the end of it
    result = fn(*args)
    if err.code != 0:
        errorClass.raise_err(err)
    return result
```

And then when you come to call an error-throwing FFI function, generate code that looks like this:

```
rust_call_with_err(ArithmeticError, _UniFFILib.arithmetic_add, a, b)
```

This avoids shared state, and by passing in the type of error as part of the call, you avoid the issue that @tarikeshaq mentions in https://github.com/mozilla/uniffi-rs/pull/215/files#r467653835 above.

Reply:
Nice one!
I hadn't thought about parallelism when I did a stateful helper, it's really a problem